### DISCUSSION TITLE
How to redirect from handle in hooks.js?

### DISCUSSION THREAD

**Discussion by ioucyf** [sveltejs/kit#6624] [Q&A]
Throw redirect(300,'/login') does not work. 

Please help.

Or is it that now the only way to do so is through +page.server.ts via the load handler for the corresponding pathname?

**Comment by mustofa-id** [comment]
Just return [`Response`](https://developer.mozilla.org/en-US/docs/Web/API/Response) object:
```js
return new Response(null, {
	status: 300,
	headers: { location: '/login' }
})
```

**Reply by ioucyf** [reply]
this doesn't work.

**Reply by teemingc** [reply]
> this doesn't work.

```ts
import type { Handle } from "@sveltejs/kit";

export const handle: Handle = ({ event, resolve }) => {
  if (event.url.pathname === "/") {
    return new Response(null, {
      status: 302,
      headers: { location: "/login" },
    });
  }

  return resolve(event);
};
```
I've tested and this works in a new project.

**Reply by ioucyf** [reply]
I'm going to try it again in a bit and try to figure out why it didn't work. 

Thank you for confirming that it does. I'll get back with updates when I do.

**Reply by ioucyf** [reply]
I just tried it again
I even made the first thing the handle function do is this:
`return new Response(null, { status: 300, headers: { location: '/user' } });` 

`hooks.js handle function` literally looks like this:
`
export const handle: Handle = function handle({ event, resolve }) {
	return new Response(null, { status: 300, headers: { location: '/user' } });
};
`

all I get is a blank page, the location bar stays to whatever i set it to.

i try going to /home (route doesn't exist) I get blank page without any redirection to /user
i try going to /user it stays there but the page is just blank
I try going to / (root) and again I get a blank page without any redirection to /user

**Reply by yousufiqbal** [reply]
I see redirect loop in your code. Wrap your response in a condition, otherwise you'll end up with **too many redirects** error in your browser.

**Reply by mustofa-id** [reply]
> `hooks.js handle function` literally looks like this: `export const handle: Handle = function handle({ event, resolve }) { return new Response(null, { status: 300, headers: { location: '/user' } }); };`

This is forever redirection. `handle` runs every time the server receives a request, you need conditionally return your custom response.  Take a look my [repo](https://github.com/mustofa-id/sveltekit-auth-hook-example/blob/4609c3870393cec4353ecbf3f6aa77d4832bada2/src/hooks/index.ts#L36) for your reference.

**Reply by ioucyf** [reply]
@yousufiqbal Thank you for your remark. I changed it to	
`if (url.pathname !== '/user') return new Response(null, { status: 307, headers: { location: '/user' } });`
to test it, and now it redirects to the correct path, but always throws the error: TypeError: **Failed to fetch dynamically imported module** instead of getting the actual page.

**Reply by ioucyf** [reply]
@mustofa-id Thank you. I will take a look.

**Reply by Antoine-lb** [reply]
It works for me but only if I put `302` and not `300`

**Comment by teemingc** [comment]
From the docs https://kit.svelte.dev/docs/modules#sveltejs-kit-redirect
> Creates a `Redirect` object. If thrown during request handling, SvelteKit will return a redirect response.

Perhaps the `redirect` helper method should be able to return a `Response` in `hooks.js` in addition to `server.js`. Might be worth opening an issue to standardise this behaviour.

**Reply by ioucyf** [reply]
> From the docs https://kit.svelte.dev/docs/modules#sveltejs-kit-redirect
> 
> > Creates a `Redirect` object. If thrown during request handling, SvelteKit will return a redirect response.

I read that, and I assumed it'd return a response that'll redirect the client to the specified location. But that does not happen in `hooks.js`.

**Comment by DrDanRyan** [comment]
I ran into the same issue a while back. I concluded that redirecting from hooks was very problematic given that hooks wasn't firing on client side navigation. My solution was to put Unauthorized errors in hooks, but to put my redirect guards into the `load` function on my layout. This works well.

**Comment by theetherGit** [comment]
I ran into same problem and i found a solution. This might not the the your required solution.
I have a CMS which have only 4 to 5 pages un protected so what i did is

```typescript
import type { Handle } from '@sveltejs/kit';

// custom redirect from joy of code `https://github.com/JoysOfCode/sveltekit-auth-cookies/blob/migration/src/hooks.ts`
function redirect(location: string, body?: string) {
	return new Response(body, {
		status: 303,
		headers: { location }
	});
}

const unProtectedRoutes: string[] = ['/', '/login']


export const handle: Handle = async ({ event, resolve }) => {
	const session = event.cookies.get('session');

	if (!session && !unProtectedRoutes.includes(event.url.pathname)) return redirect('/login', 'No authenticated user.');
	const currentUser = await db.fetch(session as string);

	if (currentUser) {
		event.locals.user = {
			name: currentUser.name,
			email: currentUser.email,
			type: currentUser.user_type,
			active: currentUser.active,
			phone: currentUser.phone
		};
	} else if (!unProtectedRoutes.includes(event.url.pathname)) {
			return redirect('/login')
		}

	return resolve(event);
};
```

you can do it like that there is bit code missing.

**Reply by suryamsj** [reply]
I followed this code, but there was a problem. If you access the /api/ route that I have created, you will be directed to the login page. So what if I don't want /api to be included?

**Reply by theetherGit** [reply]
For this scenario, I use svelte advance routing system, I create (authed) directory for managing this. In authed dir i manage all auth required pages. Code for `hooks.server.ts`

```js
import { sessionManager } from '$lib/server/redis';
import type { Handle } from '@sveltejs/kit';
import { redirect } from '@sveltejs/kit';

export const handle: Handle = async ({ event, resolve }) => {
	if (event.url.pathname.startsWith('/api')) {
		return resolve(event);
	}

	event.locals = {
		isAuthenticated: false,
		user: null
	};

	const { data, error } = await sessionManager.getSession(event.cookies);

	if (error) {
		if (event.route.id?.startsWith('/(authed)')) {
			if (event.url.pathname !== '/logout') {
				await sessionManager.deleteCookie(await event.cookies);
				throw redirect(303, `/?next=${event.url.pathname}`);
			}
		}
		return resolve(event);
	}

	if (Object.keys(data).length > 2) {
		event.locals = {
			isAuthenticated: true,
			user: data
		};
	}

	if (event.route.id?.startsWith('/(authed)')) {
		if (!event.locals.user) {
			await sessionManager.deleteCookie(await event.cookies);
			throw redirect(303, `/?next=${event.url.pathname}`);
		}
	}

	if (!event.route.id?.startsWith('/(authed)')) {
		if (event.locals.user) {
			throw redirect(303, `/dashboard/retention`);
		}
	}

	return resolve(event);
};
```

**Reply by suryamsj** [reply]
Thanks for the explanation
